### ISSUE TITLE
the release process  (what branches will be cut, when, etc)

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#60]
We could use a definition for the release process. I'd like to discuss what that process is, we'll document it afterwards. 

Below I'll outline a pretty standard release process to start the discussion:
- there is general agreement a current code version of "master" is a release candidate
- next version number is decided
- a branch, referred to as the release branch, named after the version is cut, tagged, and pushed.
- the CI system (or by hand) will deploy the build artifacts to github releases
- testers will DL/install and test:
 -- release bug fixes will get committed to the release branch and cherry-picked into master
- eventually that version is labelled a release version and user migration process begins

The git commands for the new branch and tagging are pretty straight forward:
```
git checkout -b <new_version>
git tag -a <new_version> -m '<new_version>'
git push origin <new_version>
```

**Comment by ron-murhammer** [comment]
So I think we should create a development branch so we can start merging pull requests without affecting the upcoming release. The master branch would keep the official release history while the development branch would be all changes until its merged into the master for a release. The general process is outlined here: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

The release will always be done from the latest on master and any hotfixes will be made and merged into it if needed. General development will be merged into a development branch and then merged into master when we are ready to release. We will also tag all releases so they can be kept track off.

Does that work for everyone? If so I'll create a development branch and then we can shift pull requests to that. We will then look to release and tag what is currently in master shortly as 1.8.0.6.

**Comment by DanVanAtta** [comment]
Why not the simpler feature branch workflow? 
https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

**Comment by ron-murhammer** [comment]
I guess I have a preference for keeping master clean. If we merge features directly into master then we need to be more careful about what we accept. Where as if we create a dev branch and integrate everything there first then there is less risk as even if the dev branch gets messed up we still have master to fall back to.

Though I guess we do have a small team and using a dev branch does add complexity. I guess I'm used to working with larger teams in a more enterprise setting which tends to be more cautious. I guess we could just tag releases then just merge features directly to master. 

I think I'm fine with either just not sure which people would prefer. Not sure if @veqryn has an opinion since he pretty new to the git process.

**Comment by DanVanAtta** [comment]
Master won't stay clean with merges regrettably. Git merges commits by point in time. Rebase will add commits onto something that is existing. For example, cut a branch now, commit now. Then merge that branch in a year from now. Your commit will make it to master. We may have had 30 release since then. When merged in to master, your commit will have a timestamp of when it was committed, around now, which will have been a year ago. 

What we really want I think is feature branch with squash rebase work flow. This keeps the history linear and overhead low. We don't lose anything much that the two branch gives us that I can see.

My experience with the gitflow workflow was two years on a 100 person development team 50/50 split onshore and off. It was not a good experience, it was basically the feature branch workflow plus a lot of overhead and confusion (and eventually frustration).

**Comment by DanVanAtta** [comment]
I want to point out another element of this discussion, branch age. How long branches go before being merged into master. The trick is to have "short lived branches". But the feature branch workflow (or gitflow workflow) can spawn them very easily, simply it can take a long time to get a feature finished. 

This creates an anti-pattern. Imagine a team, they create a branch, commit too it, pulling in master frequently  -  continuous integration right? Consider what happens when a second group does the same. Notice that not until both groups merge to master will either one see the others changes. Then imagine that they both finish 3 weeks later, now they've 3 weeks of work to merge.

Here is the alternative that is getting more mainstream, and that is to hide work aggressively with feature flags. It's very easy to make a menu not show up or what not, just hide the feature from the user and the rest of the code with a flag. This lets the in-development code be reviewed and merged in frequently. In the above scenarios the two groups would then integrate often whenever they pull in master, since they would also be merging back to master often (behind their feature flags, but still the code gets to be integrated and tested as one unit).

There are a number of benefits to the feature flagging, but one I'd like to point out is it helps keep turn-around time and feedback loops tight. The review is smaller, easier to do. There is less work to undo if the direction is wrong. So basically keeps the work increment small, more agile : )

Thanks for reading and the consideration. Hopefully was useful and convincing reasons for a "single master git branch model with short-lived squash-rebased feature branches that are merged via PR"

**Comment by DanVanAtta** [comment]
@ron-murhammer , @veqryn , to keep things simple and bring this topic to a close, how about we use the easier branching model for now: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

If we find we run into troubles, then we can always add the second branch and follow:
 https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

**Comment by DanVanAtta** [comment]
I think there is consensus on doing the feature-branch-workflow. I've started a how to release page here: https://github.com/triplea-game/triplea/wiki/How-To-Release

Please add it with the details when we do it for the first. We can improve the process/formatting down the road, we need got get it captured first so we can repeat the release process easily.

Closing, please re-open if there is more to discuss.

**Comment by DanVanAtta** [comment]
Re-opening to discuss: https://github.com/triplea-game/triplea/wiki/How-To-Release

@veqryn  Thanks for the initial update of the page.
I think there is a very important element that needs to be changed. Specifically updating version number as needed, tagging and branching should be the first things done, not last.

Why should it be the first thing?
- allows for master to keep going. Blocking merges to master is all sorts of bad.
- allows everyone to integrate with the new version number early.
- if we don't create the branch, and we did 2 weeks of testing already, it doesn't make sense to merge bug fixes into both master and a branch that does not yet exist!

In the same line of thought, the notes say to commit Java6 fixes before you start tagging. If we have the release branch, than you could commit to the release branch and open a PR against master. This allows us to more closely follow the rule that all commits should go through PR, without blocking someone working on a release branch.

**Comment by veqryn** [comment]
I was working on this project full time, reviewing PR's daily and merging, etc, then I would worry about it far more.  You seem to think that we've been blocking all these PR's because of master waiting for a release, when in fact neither of us have time to seriously review them (except for formatting changes).
The branching can probably be done sooner, but I want the Tag to match the actual version that goes out the door.

**Comment by DanVanAtta** [comment]
You can change tags whenever, so it can be made to match.

The argument of only time for formatting changes is not so convincing. I do recall having scripting work be redone by you rather than just accepted, IMO it's a waste of time to repeat what someone else has already done.

There is a larger conflict, a person shouldn't insist on owning a process, and then not dedicate the due time for that process. As an owner and a member of a team that creates responsibilities. I would take it as a very bad sign if you're not worried about those responsibilities, and I would take it as a even worse sign if you know you can't fulfill those responsibilities and are not looking for ways to resolve or pass them on.

The conversation about priorities on the forum boiled down to attracting more developers, and attract more players. Developers like code that is easy to modify and have their contributions be valued and see the light of day quickly. Hence an automated release process, git, Travis, it's all a priority because of that. With a low overhead and clean development process, others may be encouraged to contribute, and we'll use our time more efficiently. By that same priority, we need to be sure that this project can keep moving, even if exactly one or two people do not have time at that particular moment.

I'd like to close this comment though in acknowledgment of the years of work already put in. I was around, i saw it, the stamina to keep up for so long is pretty impressive. Though, I don't think anyone can, or is even expected to remain in such a position for so long.  Hence I'd like to see us avoid the sad situation where this project wanes simply because original contributors no longer can dedicate large amounts of time to it.
